README.md
ad-join Cookbook
============================

Library cookbook that will join an Active Directory domain


## Tested OS's

- Windows 2012R2  
- Ubuntu 14.04 (experimental)
- Ubuntu 16.04 (experimental)


## Usage

This cookbook is a library cookbook and is intended to be used by your own wrapper cookbook. See the [test/cookbooks directory](./test/cookbooks) for examples.
While the examples show running separate cookbooks for windows and linux, this isn't required. It is possible for one wrapper cookbook to manage both windows and linux hosts.


### Actions

- join
- leave

It contains a custom resource named `domain_join` with the following properties

- domain
- domain_user
- domain_password
- ou
- server (optional)
- update_hostname (optional, windows only, Set to false if you want the domain name/hostname to be different from the chef node name. (see [#5](https://github.com/NetDocuments/ad-join-cookbook/issues/5)).)
- double_reboot (optional, windows only, Will continue to reboot windows until joined to domain and breadcrumb `c:\\Windows\\chef-ad-join.txt` exists. Useful since timezone doesn't always sync after first reboot. )
- visual_warning true (optional, windows only, display a login warning to anyone who connects via RDP to the machine before chef has finished the reboots and the converge. This will override any group policy your company might have in place for displaying custom login messages.)
- hide_sensitive (optional, linux only, hide password used in realmd command, set to false for debugging)

example:  

```ruby
domain_join 'foobar' do
  domain          'example.com'
  domain_user     'binduser'
  domain_password 'correct-horse-battery-staple'
  ou              'OU=US,OU=West,OU=Web,DC=example,DC=com'
  server          'DC01'
  update_hostname true
  double_reboot true
  visual_warning true
  hide_sensitive true
  action :join
end
```

visual_warning

![](http://cl.ly/3l1I1n3X0q1G/Screenshot%202016-01-21%2012.49.45.png)


The ou must be formatted with `OU=` before each organizational unit and `DC=` before each domain component. see [test/cookbooks directory](./test/cookbooks) for an example of how to derive the OU from attributes.


## Behind the scenes

If you bootstrapped the node with the name option; e.g.

    knife bootstrap -N us-web01

Then that is the name that will be used to join the domain (not the hostname since windows randomly generates it on first boot)

**The name cannot include control characters, leading or trailing spaces, or any of the following characters: / \\ [ ].**

### Windows


In most cases, Windows hostnames must be 15 characters or less.

The cookbook creates a windows scheduled task that runs chef as soon as the VM is started. The scheduled task is deleted after all the reboots.

The cookbook will restart windows twice since some group policy objects (like the time zone) are not applied on first boot. You can change this behavior by changing the following attribute to false.

    default['ad-join']['windows']['double_reboot'] = true  

This cookbook basically runs this powershell command, then reboots

    $adminname = "EXAMPLE.COM\\bob"
    $password = 'correct-horse-battery-staple' | ConvertTo-SecureString -asPlainText -Force
    $credential = New-Object System.Management.Automation.PSCredential($adminname,$password)
    Add-computer -DomainName <EXAMPLE.COM> -OUPath <OU=FOO> -Server "<DC1.EXAMPLE.COM>'} -Credential $credential -force -Options JoinWithNewName,AccountCreate -PassThru


## Ubuntu

ad-join can join ubuntu machines to active directory. (experimental. Bug reports / pull requests encouraged)
It does not reboot or manage any of the additional files that might be required for a complete ad join

```ruby
domain_join 'foobar' do
  domain          'EXAMPLE.COM'
  domain_user     'binduser'
  domain_password 'correct-horse-battery-staple'
  ou              'OU=US,OU=West,OU=Web,DC=example,DC=com'
  server          'DC01'
  hide_sensitive true
  action :join
end
```

Common pitfalls

- Hostnames longer than 15 characters will be truncated
- NetBios names are not supported (Windows 2000 domain controllers )
- Domain is cAsE SenSITive. In most cases this needs to be all UPPERCASE.
- Debugging can be difficult, temporarily set `'hide_sensitive' false` to get additional information. domain_password will be shown in plain text.

**The ad-join cookbook is as unopinionated as possible. It will not configure `sudoers` file, `/etc/pam.d` or `/etc/krb5.conf`. Use the sudoers cookbook in your wrapper cookbook to manage those services. See [test/cookbooks/ad-join-linux directory](./test/cookbooks/ad-join-linux) for examples on how to manage those files**

This cookbook basically runs this bash command

    echo "correct-horse-battery-staple" | sudo realm join --verbose EXAMPLE.COM --user bob@EXAMPLE.COM --computer-ou OU=foobar --install=/


## Troubleshooting

### Ubuntu


```
realm: No such realm found
```

Realm is case sensitive. Try EXAMPLE.COM instead of example.com

```
realm: Not authorized to perform this action
```

Not all packages installed successfully. Verify `adcli` and `packagekit` are installed. Please open github issue if you find missing packages.

```
! Couldn't get kerberos ticket for: foo@example.com: KDC reply did not match expectations
adcli: couldn't connect to example.com domain: Couldn't get kerberos ticket for: foo@example.com: KDC reply did not match expectations
```

The domain is case sensitive. Try changing `example.com` to `EXAMPLE.COM`

```
DNS update failed: NT_STATUS_INVALID_PARAMETER
```

Make sure a fqdn is setup `hostname -f`

https://wiki.samba.org/index.php/Troubleshooting_Samba_Domain_Members

License and Authors
-------------------
Authors:  
Volodymyr Babchynskyy vbabch@softserveinc.com  
Spencer Owen sowen@netdocuments.com  


CHANGELOG.md
5.1.2
-----
Fix issue where chef 14 could complain about lack of new_resource definitions (#39)

5.1.1
-----
Fix double_reboot error on chef 14 (#37)

5.1.0
-----
Change schedule task modify to make more chef 13 friendly

5.0.4
-----
Fix typo in metadata

5.0.3
-----

Work around bug with chef 13
https://github.com/chef/chef/issues/6824#issuecomment-367141438

-----
Fix leave action on chef 13 (#31)

5.0.1
-----
Fix scheduled task not running on windows

5.0.0
-----
Adds Ubuntu support  
Fix Chef 13. Requires 13.4.19 or greater (or >=12.7) (#12, #20, #23)

4.12.1
------
Throws error if running on chef 11 or chef 13
Temporary fix until this issue is fixed https://github.com/NetDocuments/ad-join-cookbook/issues/23

4.12.0
------
Fixes issue #19
Fixes deprecation warning for chef 13

4.11.1
------
Fix berkshelf supermarket url

4.11.0
------
Abort if hostname is longer than 15 characters on windows

4.10.0
------
Adds domain leave functionality (#16 metalseargolid)

4.9.0
-----
Fix: Scheduled task wont run if time zone changes on reboot (#13)

4.8.0
-----
Fix: No longer gives deprecation warnings if 'server' is nil. (#9)

4.7.0
-----
Improvement: Adds name to scheduled task, removing need for workaround http://bit.ly/1WDZ1kn
Change: Changes c:\\windows\\chef-ad-join.txt to windows friendly path c:/windows/chef-ad-join.txt

4.6.1
-----
Fix: Warning registry key not cleaned up

4.6.0
-----
Add: 'server' parameter to allow for specifying a specific domain controller
Fix: Warning message wouldn't be displayed (#4)

4.5.0
-----
Fix: Passwords with special characters now work properly (#7 Thanks opsline-radek)
Fix: OU Parameter is now truly optional (#6 Thanks opsline-radek)

4.4.0
-----
Adds new attribute default['ad-join']['windows']['update_hostname']

4.3.0
-----
Adds warning attribute

4.2.0
-----
Fixes incorrect CWD in sched task (issue #3)
Fixes incorrect ohai fact "node['os']"

4.1.0
-----
Fixes powershell error when vm name is same as bootstrap name.  issue #2

4.0.2
-----
Updates metadata for supermarket

4.0.1
-----
Fixes attribute name for double reboot

4.0.0
-----
Created new git repo for public release on github

3.0.2
-----
Create breadcrumb only if missing

3.0.1
-----
Fixes OU not having quotes

3.0.0
-----
Complete rewrite to make it a library cookbook

2.0.2
-----
More verbose logging in scheduled task

2.0.1
-----
Reduces timeout to 30 seconds

1.0.0
-----
general cleanup, removed private domain name and so on, prepared for public release

0.9.0
-----
removed private usernames and passwords

0.8.1
-----
rubocop convention alerts accepted

0.8.0
-----
changed databag name

0.6.1
-----
rubocop check for line length now is 120 symbols

0.6.0
-----
rubocop and foodcritic inspections added

0.5.1
-----
icon added

0.5.0
-----
tests added

0.4.0
-----
Added possibility to run it on teamcity CI

0.3.0
-----
Fixed, directory server is unavailable issue, code commented for future use

0.2.0
-----
Passwords moved into databag

0.1.1
-----
added ohai reload for new fqdn resolution in chef

0.1.0
-----
Initial release of ad-join


resources/domain_join.rb
resource_name :domain_join
property :domain, String, required: true
property :domain_user, String, required: true
property :domain_password, String, required: true
property :ou, [String, NilClass], required: false, default: nil
property :server, [String, NilClass], required: false, default: nil
property :update_hostname, [TrueClass, FalseClass], required: false, default: true
property :double_reboot, [TrueClass, FalseClass], required: false, default: true
property :visual_warning, [TrueClass, FalseClass], required: false, default: true
property :hide_sensitive, [TrueClass, FalseClass], required: false, default: true

default_action :join

Chef::Log.warn( "node['ad-join']['windows']['update_hostname'] deprecated") if !!(node['ad-join']['windows']['update_hostname'])
Chef::Log.warn( "node['ad-join']['windows']['visual_warning'] deprecated") if !!(node['ad-join']['windows']['visual_warning'])
Chef::Log.warn( "node['ad-join']['windows']['update_hostname'] deprecated") if !!(node['ad-join']['windows']['visual_warning'])

# ohai domain attributes
# node['kernel']['cs_info']['domain_role']

# StandaloneWorkstation = 0
# MemberWorkstation = 1
# StandaloneServer = 2
# MemberServer = 3
# BackupDomainController = 4
# PrimaryDomainController = 5

action :join do
  # Set the computer name to the same name provided by -N parameter in  knife boostrap -N 'node01'
  if Chef::Config[:node_name] != node['hostname'] && Chef::Config[:node_name] != node['fqdn'] && update_hostname == true
    # Abort if hostname is more than 15 characters long on windows
    raise if Chef::Config[:node_name].length > 15

    newcomputername = Chef::Config[:node_name]
    # renew info about nodename on chef server after reload
    ohai 'reload' do
      action :reload
    end
  else
    newcomputername = node['hostname']
  end

  reboot 'Restart Computer' do
    action :nothing
  end

  case node['os']
  when 'windows'
    warning_caption = 'Chef is joining the domain'
    warning_text = 'The chef cookbook ad-join is currently in the middle of joining the domain, and the server is about to be restarted'
    warning_key = 'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System'

    # Display a warning incase anyone remotes into the machine before all reboots have finished
    # http://www.techrepublic.com/blog/windows-and-office/adding-messages-to-windows-7s-logon-screen/
    registry_key 'warning' do
      key warning_key
      values [
        { name: 'legalnoticecaption', type: :string, data: warning_caption },
        { name: 'legalnoticetext', type: :string, data: warning_text }
      ]
      only_if { visual_warning == true }
      action :nothing
    end

    # Installs task for chef-client run after reboot, needed for ohai reload
    windows_task 'chef ad-join' do
      task_name 'chef ad-join' # http://bit.ly/1WDZ1kn
      user 'SYSTEM'
      command 'chef-client -L C:\chef\chef-ad-join.log'
      run_level :highest
      frequency :onstart
      only_if { node['kernel']['cs_info']['domain_role'].to_i == 0 || node['kernel']['cs_info']['domain_role'].to_i == 2 }
      notifies :create, 'registry_key[warning]', :immediately # http://bit.ly/1WDZ1kn
      action :create
    end

    # Modify the start time to make sure GP doesn't set task into future
    # https://github.com/NetDocuments/ad-join-cookbook/issues/13
    # schedtask.exe won't allow this to be combined with task creation
    windows_task 'modify sched task start time' do
      task_name 'chef ad-join' # http://bit.ly/1WDZ1kn
      start_day '06/09/2016'
      start_time '01:00'
      only_if { node['kernel']['cs_info']['domain_role'].to_i == 0 || node['kernel']['cs_info']['domain_role'].to_i == 2 }
      action :change
    end

    powershell_script 'ad-join' do
      code <<-EOH
      $adminname = "#{new_resource.domain}\\#{domain_user}"
      $password = '#{domain_password}' | ConvertTo-SecureString -asPlainText -Force
      $credential = New-Object System.Management.Automation.PSCredential($adminname,$password)

      if ( '#{newcomputername}' -eq $(hostname) ) {
        Write-Host "Skipping computer rename since already named: #{newcomputername}"
      }
      else {
        Write-Host "Renaming computer from $($hostname) to #{newcomputername}"
        Rename-Computer -NewName '#{newcomputername}'
      }
      sleep 5
      Add-computer -DomainName #{new_resource.domain} #{ou.nil? ? '' : '-OUPath "' + ou + '"'} #{server.nil? ? '' : '-Server "' + server + '"'} -Credential $credential -force -Options JoinWithNewName,AccountCreate -PassThru #-Restart

      # Old way, somtimes Domain controller busy error occured
      # Add-Computer  #{newcomputername} -DomainName #{new_resource.domain} -OUPath #{ou} -Credential $credential -Restart -PassThru
      # Add-Computer -ComputerName Server01 -LocalCredential Server01\Admin01 -DomainName Domain02 -Credential Domain02\Admin02 -Restart -Force
      EOH
      only_if { node['kernel']['cs_info']['domain_role'].to_i == 0 || node['kernel']['cs_info']['domain_role'].to_i == 2 }
      notifies :reboot_now, 'reboot[Restart Computer]', :immediately
    end

    # Reboot the computer a second time
    # Needed on windows systems to apply some group policy objects (like timezone)
    file 'c:/Windows/chef-ad-join.txt' do
      content 'Placed by ad-join cookbook. Cookbook will keep rebooting windows until server is part of a domain and this file exists. DONT DELETE'
      action :create_if_missing
      only_if { double_reboot == true }
      notifies :reboot_now, 'reboot[Restart Computer]', :immediately
    end

    windows_task 'chef ad-join' do
      notifies :delete, 'registry_key[warning]', :delayed
      action :delete
    end

  when 'linux'

    case node['platform_version']
    when '16.04'

      #TODO: don't run every time, just first time
      apt_update 'update' do
        ignore_failure true
        action :nothing
      end

      # AD Join loosely based on this document https://help.ubuntu.com/lts/serverguide/sssd-ad.html
      # https://tutel.me/c/unix/questions/256626/sssd+realm+discover+not+authorized+to+perform+this+action
      %w(realmd sssd-tools sssd libnss-sss libpam-sss adcli packagekit).each do |pkg|
        package pkg do
          action :install
          notifies :update, 'apt_update[update]', :before
        end
      end

      # https://answers.launchpad.net/ubuntu/+question/293540
      execute 'realm join' do
        environment 'DOMAIN_PASS' => domain_password
        command <<-EOH
          echo "${DOMAIN_PASS}" | sudo realm join --verbose #{new_resource.domain} --user #{domain_user}@#{new_resource.domain} --computer-ou #{ou} --install=/
          EOH
        not_if <<-EOH
          domain=$(sudo realm list -n| tr '[:upper:]' '[:lower:]');
          # echo domain is ${domain} > /tmp/domain;
          # echo "resource domain is #{new_resource.domain.downcase}" >> /tmp/domain;
          if [ "${domain}" != "#{new_resource.domain.downcase}" ]; then
            # echo "${domain} doesnt match #{new_resource.domain.downcase}" >> /tmp/domain;
            exit 1;
          else
            # echo "${domain} matches #{new_resource.domain.downcase}" >> /tmp/domain;
            exit 0;
          fi
          EOH
        sensitive true if hide_sensitive == true
      end

    else
      Chef::Log.fatal("Only Ubuntu 16.04 currently supported. Found: #{node['platform']} #{node['platform_version']}")
    end
  else
    Chef::Log.fatal("Platform: #{node['platform']} not supported")
  end
end

action :leave do
  reboot 'Restart Computer' do
    action :nothing
  end

  case node['os']
  when 'windows'
    warning_caption = 'Chef is leaving the domain'
    warning_text = 'The chef cookbook ad-join is currently in the middle of leaving the domain, and the server is about to be restarted'
    warning_key = 'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System'

    # Display a warning incase anyone remotes into the machine before all reboots have finished
    # http://www.techrepublic.com/blog/windows-and-office/adding-messages-to-windows-7s-logon-screen/
    registry_key 'warning' do
      key warning_key
      values [
        { name: 'legalnoticecaption', type: :string, data: warning_caption },
        { name: 'legalnoticetext', type: :string, data: warning_text }
      ]
      only_if { node['ad-join']['windows']['visual_warning'] == true }
      action :nothing
    end

    # Installs task for chef-client run after reboot, needed for ohai reload
    windows_task 'chef ad-join leave' do
      task_name 'chef ad-join leave' # http://bit.ly/1WDZ1kn
      user 'SYSTEM'
      command 'chef-client -L C:\chef\chef-ad-join.log'
      run_level :highest
      frequency :onstart
      only_if { node['kernel']['cs_info']['domain_role'].to_i == 1 || node['kernel']['cs_info']['domain_role'].to_i == 3 }
      notifies :create, 'registry_key[warning]', :immediately # http://bit.ly/1WDZ1kn
      action :create
    end

    # Modify the start time to make sure GP doesn't set task into future
    # https://github.com/NetDocuments/ad-join-cookbook/issues/13
    # schedtask.exe won't allow this to be combined with task creation
    windows_task 'chef ad-join leave start time' do
      task_name 'chef ad-join leave' # http://bit.ly/1WDZ1kn
      start_day '06/09/2016'
      start_time '01:00'
      only_if { node['kernel']['cs_info']['domain_role'].to_i == 1 || node['kernel']['cs_info']['domain_role'].to_i == 3 }
      action :change
    end

    powershell_script 'ad-join-leave' do
      code <<-EOH
      $adminname = "#{domain}\\#{domain_user}"
      $password = '#{domain_password}' | ConvertTo-SecureString -asPlainText -Force
      $credential = New-Object System.Management.Automation.PSCredential($adminname,$password)

      Remove-Computer -UnjoinDomainCredential $credential -Force -PassThru
      EOH
      only_if { node['kernel']['cs_info']['domain_role'].to_i == 1 || node['kernel']['cs_info']['domain_role'].to_i == 3 }
      notifies :reboot_now, 'reboot[Restart Computer]', :immediately
    end

    file 'C:/Windows/chef-ad-join.txt' do
      action :delete
    end

    windows_task 'chef ad-join task delete' do
      task_name 'chef ad-join leave' # http://bit.ly/1WDZ1kn
      notifies :delete, 'registry_key[warning]', :delayed
      action :delete
    end

  when 'linux'
    raise "Linux can't yet unjoin from domain"

  else
    Chef::Log.fatal("Platform: #{node['platform']} not supported")
  end
end
